### PR TITLE
feat(core): ldml marker normalization  🙀

### PR DESCRIPTION
**Pull request by srl295**
- [X] add some failing tests
- [x] fix them

For: #9468

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by srl295** [comment]
moved the failing bengali test case here for fixin'!

**Comment by mcdurdin** [comment]
I think we need to bring up to date with master in order for some of the Linux builds to work.

File: core/tests/unit/ldml/test_transforms.cpp [hunk 1]
```diff
@@ -427,10 +428,196 @@ int test_map() {
...
+
+  {
+    marker_map map;
+    std::cout << __FILE__ << ":" << __LINE__ << "   - noop test w markers" << std::endl;
+    const std::u32string src =
+        U"6\U0000ffff\U00000008\U00000001e\U0000ffff\U00000008\U00000002\U00000320\U0000ffff\U00000008\U00000003\U00000300"
+        U"\U0000ffff\U00000008\U00000004";
+    const std::u32string expect =
+        U"6\U0000ffff\U00000008\U00000001e\U0000ffff\U00000008\U00000002\U00000320\U0000ffff\U00000008\U00000003\U00000300"
+        U"\U0000ffff\U00000008\U00000004";
```
Comment (mcdurdin) [inline core/tests/unit/ldml/test_transforms.cpp:550]:
```suggestion
    const std::u32string expect = src;
```

?

File: core/tests/unit/ldml/test_transforms.cpp [hunk 2]
```diff
@@ -427,10 +428,196 @@ int test_map() {
...
+    assert_equal(map[MARKER_BEFORE_EOT], 0x4L);
+  }
+
+  {
+    marker_map map;
+    std::cout << __FILE__ << ":" << __LINE__ << "   - complex test" << std::endl;
+    const std::u32string src = // already in order: 320+300
+        U"6\U0000ffff\U00000008\U00000001e\U0000ffff\U00000008\U00000002\U00000320\U0000ffff\U00000008\U00000003\U00000300\U0000ffff\U00000008\U00000004";
+    const std::u32string expect =
+        U"6\U0000ffff\U00000008\U00000001e\U0000ffff\U00000008\U00000002\U00000320\U0000ffff\U00000008\U00000003\U00000300\U0000ffff\U00000008\U00000004";
```
Comment (mcdurdin) [inline core/tests/unit/ldml/test_transforms.cpp:567]:
```suggestion
    const std::u32string expect = src;
```

**Review by mcdurdin — COMMENTED** [review]
Oh my C++ Unicode escapes are _so_ hard to read. Suggestions added

**Comment by mcdurdin** [comment]
#9761 had 4 failures, Windows x 2 were:

```
02:37:59   ../../../src/debuglog.cpp(467): error C2220: the following warning is treated as an error
02:37:59   ../../../src/debuglog.cpp(467): warning C4477: 'snprintf' : format string '%4.6X' requires an argument of type 'unsigned int', but variadic argument 1 has type 'char32_t'
```

https://build.palaso.org/buildConfiguration/KeymanDesktop_TestPullRequests/422546?buildTab=log&focusLine=2451&linesState=370&logView=flowAware

and macOS:

```
73/83 keyman_core:ldml-keyboards / k_008_transform_norm        FAIL            0.06s   killed by signal 6 SIGABRT
...
83/83 keyman_core:ldml / test_transforms                       FAIL            0.81s   killed by signal 6 SIGABRT
```

https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_macOS/422542?buildTab=log&focusLine=3094&linesState=89&logView=flowAware

Developer build failure was codesigning fail. Will restart that one but it will probably fail on Core?

**Comment by srl295** [comment]
Hm ok Thanks

**Comment by srl295** [comment]
put in a patch

**Comment by srl295** [comment]
@mcdurdin fail didn't seem to repro here, but it turns out:

image

but 

```
../../../tests/unit/ldml/ldml_test_source.cpp:735: warning: (!rep_tests.empty()) && (rep_tests.size() > 0)
\x1b[38;5;196mTest failed with 0 at ../../../tests/unit/ldml/ldml.cpp:192:\x1b[0m
  ci->type != KM_CORE_CT_END && test_ci != test_context.end()
```

Seems like the status isn't being reported properly. So i'm getting false positives on my end.

**Comment by srl295** [comment]
@ermshiperete @mcdurdin  `std::exit(0)` was being called on test failure… take a look at the test_assert.h change

**Comment by mcdurdin** [comment]
So, still some [failing tests](https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_macOS/422852?buildTab=log&focusLine=3115&linesState=110&logView=flowAware)?

```
75/83 keyman_core:ldml-keyboards / k_008_transform_norm        FAIL            0.07s   killed by signal 6 SIGABRT
...
83/83 keyman_core:ldml / test_transforms                       FAIL            0.83s   killed by signal 6 SIGABRT
```

**Comment by mcdurdin** [comment]
> @ermshiperete @mcdurdin `std::exit(0)` was being called on test failure… take a look at the test_assert.h change

How? the exit code was `100*__LINE__` or `100*__LINE__+(result)` and `__LINE__` should never be 0?

It seems this change to test_assert.h reduces the information provided in case of test failure, because we no longer have a failing line reported.

**Comment by srl295** [comment]
I cleaned this up and moved the test-case-fix stuff into another PR

**Comment by mcdurdin** [comment]
[Keyman Core macOS build](https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_macOS/424259?buildTab=log&focusLine=3263&linesState=99&logView=flowAware) failed with:

```
Assertion failed: (*i == LDML_MARKER_CODE), function remove_markers, file ldml_transforms.cpp, line 1007.
```

**Comment by srl295** [comment]
doesn't repro locally, checking still

**Comment by srl295** [comment]
> doesn't repro locally, checking still

fixed, it was actually a bad assert.  The case it was asserting for was handled in error checking code one line below. I had a local build configuration that ended up ignoring the assert!

File: core/tests/unit/ldml/ldml.cpp [hunk 3]
```diff
@@ -118,11 +122,23 @@ apply_action(
...
-      assert(context.back().type == KM_CORE_CT_CHAR);
-      assert(context.back().character == ch);
-      context.pop_back();
+      if (act.backspace.expected_type == KM_CORE_BT_CHAR) {
+        assert(ch == act.backspace.expected_value);
+        assert(context.back().type == KM_CORE_CT_CHAR);
+        assert(context.back().character == ch);
+        context.pop_back();
+      } else {
+        assert(act.backspace.expected_type == KM_CORE_BT_UNKNOWN); // else it must be unknown
```
Comment (mcdurdin) [inline core/tests/unit/ldml/ldml.cpp:131]: BT_UNKNOWN should only be used when the context is empty.

**Review by mcdurdin — APPROVED** [review]
LGTM as far as it goes -- still more iterations needed on multiple markers between two chars I guess.

One issue possible with BT_UNKNOWN type backspaces -- it should only happen with an empty context.

File: core/tests/unit/ldml/ldml.cpp (hunk 3 above)
Comment (srl295) [reply core/tests/unit/ldml/ldml.cpp]: added an assert for that request

**Comment by srl295** [comment]
might as well update. mac failure seemed to be timeout

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.215-alpha](https://keyman.com/downloads/releases/alpha/17.0.215)

**Pull request merged**